### PR TITLE
Validation of the VolatileDB

### PR DESCRIPTION
**Pull request by kderme**
Addresses part 2 of https://github.com/input-output-hk/ouroboros-network/issues/1304#issuecomment-581289381 and more:

- there is an option to check if each parsed block is valid, while parsing.
- the db now recovers when there is a duplicated slot and ignores invalid filenames.
- added tests for this.
- added traces.

`readIncrementalOffsets` is no longer used anywhere. Worth keeping?

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs [hunk 1]
```diff
@@ -418,10 +429,11 @@ tryCollectFile :: forall m h blockId
                -> InternalState blockId h
                -> (FileId, FileInfo blockId)
                -> m (InternalState blockId h)
-tryCollectFile hasFS@HasFS{..} env slot st@InternalState{..} (fileId, fileInfo) =
+tryCollectFile hasFS env@VolatileDBEnv{..} slot st@InternalState{..}
+  (fileId, fileInfo) =
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: I would revert this change (so it's one line again) and use the following instead:

```haskell
tryCollectFile ...
    ....
  where
    VolatileDBEnv { _dbErr, _tracer } = env
```

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/VolDB.hs [hunk 2]
```diff
@@ -494,22 +514,36 @@ blockFileParser' :: forall m blk h. (IOLike m, HasHeader blk)
...
                  -> (forall s. Decoder s (Lazy.ByteString -> blk))
+                 -> (blk -> Bool)
                  -> VolDB.Parser
-                     Util.CBOR.ReadIncrementalErr
+                     (BlockFileParserError (HeaderHash blk))
                      m
                      (HeaderHash blk)
-blockFileParser' hasFS isEBB encodeBlock decodeBlock = VolDB.Parser $
-       fmap (fmap (fmap fst)) -- Drop the offset of the error
-     . Util.CBOR.readIncrementalOffsets hasFS decoder'
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/VolDB.hs]: Yeah, if `readIncrementalOffsets` is no longer used, I would remove it. We can always resurrect it if necessary.

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/VolDB.hs [hunk 3]
```diff
@@ -494,22 +514,36 @@ blockFileParser' :: forall m blk h. (IOLike m, HasHeader blk)
...
+                    m
+                    (Maybe (Util.CBOR.ReadIncrementalErr, Word64))
+                 -> m (VolDB.ParsedInfo (HeaderHash blk),
+                    Maybe (BlockFileParserError (HeaderHash blk)))
+    checkEntries = go []
+      where
+        go parsed stream = S.next stream >>= \case
+          Left mbErr -> return (reverse parsed, BlockReadErr . fst <$> mbErr)
+          Right ((offset, (size, blk)), stream') | checkIntegrity blk ->
+            let newParsed = (offset, (VolDB.BlockSize size, extractInfo' blk))
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/VolDB.hs]: We have to be careful here to not accidentally hold on to the block. As we're returning a tuple, we can't rely on the data type to do the forcing for us. You can do the following:
```haskell
let !blockInfo = extractInfo' blk
    newParsed  = (offset, (VolDB.BlockSize size, blockInfo))
...
```

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/VolDB.hs [hunk 4]
```diff
@@ -494,22 +514,36 @@ blockFileParser' :: forall m blk h. (IOLike m, HasHeader blk)
...
+            in go (newParsed : parsed) stream'
+          Right ((_, (_, blk)), _) -> return (reverse parsed,
+            Just $ BlockCorruptedErr $ VolDB.bbid $ extractInfo' blk)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/VolDB.hs]: Same here:
```haskell
Right ((_, (_, blk)), _) ->
  let !bid = VolDB.bbid $ extractInfo' blk
  in return (reverse parsed, Just (BlockCorruptedErr bid))
```

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/VolDB.hs [hunk 5]
```diff
@@ -475,17 +490,22 @@ getBlockComponent db blockComponent hash = withDB db $ \vol ->
...
 
 {-------------------------------------------------------------------------------
-  Auxiliary: parsing
+  Parsing
 -------------------------------------------------------------------------------}
 
+data BlockFileParserError hash =
+    BlockReadErr Util.CBOR.ReadIncrementalErr
+  | BlockCorruptedErr hash
+  deriving (Eq, Show)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/VolDB.hs]: Export this type too.

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/StateMachine.hs [hunk 6]
```diff
@@ -381,6 +387,10 @@ preconditionImpl Model{..} (At (CmdErr cmd err)) =
         case fmap fst . snd <$> M.lookup file (index dbModel) of
           Nothing   -> Bot
           Just bids -> isOpen .&& (thHash $ testHeader tb) `elem` bids
+      PutCorrupted tb        -> isOpen .&& Boolean (and
+        [ (not $ checkBlockIntegrity tb)
+        , (not $ M.member (thHash $ testHeader tb) (mp dbModel))
+        , (slotAfterGC $ Just $ thSlotNo $ testHeader tb) ])
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/StateMachine.hs]: I would try to use [`Logic`](https://hackage.haskell.org/package/quickcheck-state-machine-0.6.0/docs/Test-StateMachine-Logic.html) as much as possible instead of `Bool`, so you get a more precise message in case it no longer holds. (Also: use `M.notMember` to avoid the extra `Not`)

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/StateMachine.hs [hunk 7]
```diff
@@ -609,6 +638,11 @@ mockImpl model cmdErr = At <$> return mockResp
   where
     (mockResp, _dbModel') = step model cmdErr
 
+checkBlockIntegrity :: TestBlock -> Bool
+checkBlockIntegrity testBlock =
+  hashHeader (testHeader testBlock) == thHash (testHeader testBlock)
+  || not (tbIsValid (testBody testBlock))
+
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/StateMachine.hs]: Isn't this the same as `Test.Ouroboros.Storage.TestBlock.testBlockIsValid`?

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs (hunk 1 above)
Comment (kderme) [reply ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs]: This is not so easy because `VolatileDBEnv` is Existential type, but I did the same for `InternalState` to keep it a single like.

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/VolDB.hs [hunk 8]
```diff
@@ -494,22 +516,38 @@ blockFileParser' :: forall m blk h. (IOLike m, HasHeader blk)
...
+    checkEntries :: Stream (Of (Word64, (Word64, blk)))
+                    m
+                    (Maybe (Util.CBOR.ReadIncrementalErr, Word64))
+                 -> m (VolDB.ParsedInfo (HeaderHash blk),
+                    Maybe (BlockFileParserError (HeaderHash blk)))
+    checkEntries = go []
+      where
+        go parsed stream = S.next stream >>= \case
+          Left mbErr -> return (reverse parsed, BlockReadErr . fst <$> mbErr)
+          Right ((offset, (size, blk)), stream') | checkIntegrity blk ->
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/VolDB.hs]: This check is fairly expensive and will slow down startup. Can you make the `checkIntegrity` check optional? Just like we have the two validation policies for the ImmutableDB. By default, we won't do the check, we'll only do it when started in recovery mode. The validation policy should be an argument to the VolatileDB.

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/StateMachine.hs [hunk 9]
```diff
@@ -157,6 +158,7 @@ data Cmd
     | CreateFile
     | CreateInvalidFile
     | DuplicateBlock FsPath TestBlock
+    | PutCorrupted TestBlock
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/StateMachine.hs]: How is this different from `Corrupt`? To me `Corrupt` sounds more general and thus preferable.

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/VolDB.hs (hunk 8 above)
Comment (kderme) [reply ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/VolDB.hs]: I missed this comment, don't review yet...

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/StateMachine.hs (hunk 7 above)
Comment (kderme) [reply ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/StateMachine.hs]: Almost yes, this also checkes `tbIsValid`

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 10]
```diff
@@ -141,9 +143,12 @@ run tracers protocolTracers chainDbTracer diffusionTracers diffusionArguments
               -- When the last shutdown was not clean, validate the complete
               -- ChainDB to detect and recover from any corruptions. This will
               -- override the default value /and/ the user-customised value of
-              -- the 'ChainDB.cdbValidation' field.
+              -- the 'ChainDB.cdbValidation' and the
+              -- 'ChainDB.cdbBlockValidation' fields.
             = (customiseChainDbArgs args)
-              { ChainDB.cdbValidation = ValidateAllEpochs }
+              { ChainDB.cdbValidation      = ValidateAllEpochs
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]:
```suggestion
              { ChainDB.cdbImmValidation      = ValidateAllEpochs
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 11]
```diff
@@ -141,9 +143,12 @@ run tracers protocolTracers chainDbTracer diffusionTracers diffusionArguments
...
+              , ChainDB.cdbBlockValidation = ValidateAll
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]:
```suggestion
              , ChainDB.cdbVolValidation = ValidateAll
```

**Comment by mrBliss** [comment]
This ready for merging. Thanks @kderme for the validation code. I cleaned up some bits and implemented part 1 of https://github.com/input-output-hk/ouroboros-network/issues/1304#issuecomment-581289381.
